In [1]:
#Library import
import warnings
import os
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
warnings.filterwarnings('ignore')

In [3]:
# OS로 디렉토리를 설정한 후 데이터를 읽어들인다.
#os.chdir("Ans1")

train_xdata = np.loadtxt("titanic_3355/train_xdata.csv",delimiter=',', skiprows=1)
train_label = np.loadtxt("titanic_3355/train_label.csv",delimiter=',',skiprows=1)
test_xdata = np.loadtxt("titanic_3355/test_xdata.csv",delimiter=',',skiprows=1)
test_label = np.loadtxt("titanic_3355/test_label.csv",delimiter=',',skiprows=1)

In [4]:
# Standard Scaling 진행 
ss = StandardScaler()
train_xdata = ss.fit_transform(train_xdata) #train_xdata를 Scaling 한다.
test_xdata = ss.transform(test_xdata)


In [5]:
#모델의 Accuracy 반환 (DONT TOUCH)
def returnACC(model):
    model.fit(train_xdata, train_label)
    pred = model.predict(test_xdata)
    return accuracy_score(test_label, pred)

rf =RandomForestClassifier(random_state=3355)
dt = DecisionTreeClassifier(random_state=3355)
lr = LogisticRegression(random_state=3355)
print('DT : ', returnACC(dt),'\nRF : ',returnACC(rf),'\nLR :', returnACC(lr))

def grid_search(model, parameters):
    grid_dclf = GridSearchCV(model, param_grid=parameters, scoring='accuracy', cv=5)
    grid_dclf.fit(train_xdata, train_label)
    print(accuracy_score(grid_dclf.best_estimator_.predict(test_xdata), test_label))
    print(grid_dclf.best_params_)

DT :  0.7870813397129187 
RF :  0.8181818181818182 
LR : 0.937799043062201


In [8]:
# 모델의 파라미터 조절하기! 
# decision Tree의 성능을 높이고 RandomForest의 성능의 변화를 확인한다.
rf =RandomForestClassifier(random_state=3355)
dt = DecisionTreeClassifier(random_state=3355)
lr = LogisticRegression(random_state=3355)

parameters={'max_depth':[1,2,3,4,5],'min_samples_leaf':[2,3,4,5],'min_samples_split':[3,4,5,6]} #파라미터를 튜닝하세요.

grid_search(dt, parameters)

0.8827751196172249
{'max_depth': 5, 'min_samples_leaf': 3, 'min_samples_split': 3}


In [16]:
parameters={'max_depth':[1,2,3,4,5,6],'min_samples_leaf':[2,3],'min_samples_split':[4,5,6,7]} #파라미터를 튜닝하세요.

grid_search(rf, parameters)

0.8588516746411483
{'max_depth': 6, 'min_samples_leaf': 2, 'min_samples_split': 6}


In [ ]:
#Q1 : Decision Treed의 Accuracy를 0.83 이상으로 높였을 때 파라미터는 무엇인가요? 
#Q2 : Random Forest로 Accuracy를 0.86 이상으로 높일 수 있나요? (T/F)

### A1: dt의accuracy가 0.88일 때, 파라미터가 {'max_depth': 5, 'min_samples_leaf': 3, 'min_samples_split': 3} 이다.
### A2: F. 여러번 튜닝을 시도하였는데 파라미터가 저럴때 가장 높은 accuracy가 나왔음.